staging/src/k8s.io/dynamic-resource-allocation/api/v1beta1/conversion.go

```diff
@@ -0,0 +1,274 @@
+		}
+		exactDeviceRequest.AllocationMode = resourceapi.DeviceAllocationMode(in.AllocationMode)
+		exactDeviceRequest.Count = in.Count
+		exactDeviceRequest.AdminAccess = in.AdminAccess
+		var tolerations []resourceapi.DeviceToleration
+		for _, e := range in.Tolerations {
+			var toleration resourceapi.DeviceToleration
+			if err := Convert_v1beta1_DeviceToleration_To_v1beta2_DeviceToleration(&e, &toleration, s); err != nil {
+				return err
+			}
+			tolerations = append(tolerations, toleration)
+		}
+		exactDeviceRequest.Tolerations = tolerations
+		out.Exactly = &exactDeviceRequest
+	}
+	return nil
+}
+
+func hasAnyMainRequestFieldsSet(deviceRequest *resourcev1beta1.DeviceRequest) bool {
+	return deviceRequest.DeviceClassName != "" ||
+		deviceRequest.Selectors != nil ||
+		deviceRequest.AllocationMode != "" ||
+		deviceRequest.Count != 0 ||
+		deviceRequest.AdminAccess != nil ||
+		deviceRequest.Tolerations != nil
+}
+
+func Convert_v1beta2_DeviceRequest_To_v1beta1_DeviceRequest(in *resourceapi.DeviceRequest, out *resourcev1beta1.DeviceRequest, s conversion.Scope) error {
+	if err := autoConvert_v1beta2_DeviceRequest_To_v1beta1_DeviceRequest(in, out, s); err != nil {
+		return err
+	}
+	if in.Exactly != nil {
+		out.DeviceClassName = in.Exactly.DeviceClassName
+		if in.Exactly.Selectors != nil {
+			selectors := make([]resourcev1beta1.DeviceSelector, 0, len(in.Exactly.Selectors))
+			for i := range in.Exactly.Selectors {
+				var selector resourcev1beta1.DeviceSelector
+				err := Convert_v1beta2_DeviceSelector_To_v1beta1_DeviceSelector(&in.Exactly.Selectors[i], &selector, s)
+				if err != nil {
+					return err
+				}
+				selectors = append(selectors, selector)
+			}
+			out.Selectors = selectors
+		}
+		out.AllocationMode = resourcev1beta1.DeviceAllocationMode(in.Exactly.AllocationMode)
+		out.Count = in.Exactly.Count
+		out.AdminAccess = in.Exactly.AdminAccess
+		var tolerations []resourcev1beta1.DeviceToleration
+		for _, e := range in.Exactly.Tolerations {
+			var toleration resourcev1beta1.DeviceToleration
+			if err := Convert_v1beta2_DeviceToleration_To_v1beta1_DeviceToleration(&e, &toleration, s); err != nil {
+				return err
+			}
+			tolerations = append(tolerations, toleration)
+		}
+		out.Tolerations = tolerations
+	}
+	return nil
+}
+
+func Convert_v1beta1_ResourceSliceSpec_To_v1beta2_ResourceSliceSpec(in *resourcev1beta1.ResourceSliceSpec, out *resourceapi.ResourceSliceSpec, s conversion.Scope) error {
+	if err := autoConvert_v1beta1_ResourceSliceSpec_To_v1beta2_ResourceSliceSpec(in, out, s); err != nil {
+		return err
+	}
+	if in.NodeName == "" {
+		out.NodeName = nil
+	} else {
+		out.NodeName = &in.NodeName
+	}
+	if !in.AllNodes {
+		out.AllNodes = nil
+	} else {
+		out.AllNodes = &in.AllNodes
+	}
+	return nil
+}
+
+func Convert_v1beta2_ResourceSliceSpec_To_v1beta1_ResourceSliceSpec(in *resourceapi.ResourceSliceSpec, out *resourcev1beta1.ResourceSliceSpec, s conversion.Scope) error {
```

Comment:
Just to clarify, this v1beta2-to-v1beta1 conversion is handled systematically by the API server.  
This is for the case when objects are created using the v1beta2 type, but some older plugins still query using the v1beta1 type.

Reply:
The apiserver does a similar conversion to support different clients where one uses v1beta1 and the other uses v1beta2. But it does that under different API endpoints (`/apis/resource.k8s.io/v1beta1/resourceslices` vs `/apis/resource.k8s.io/v1beta2/resourceslices`).

What we want is to let the DRA driver authors use the v1beta2 API, convert on the client side and use the v1beta1 endpoint.

Reply:
sorry but I miss context, why do we want to force the use of the v1beta1 endpoint?

Reply:
Compatibility with Kubernetes 1.32: that one only supports v1beta1, not v1beta2.

This way, a DRA driver developed with the helper code from v1.33 is compatible with any Kubernetes release >= v1.32.